File: prog_kotlin/network_plann/network_plann.kt
package network_plann

//Вычисление предшественников
private fun calcualte_previous(matrix: ArrayList<ArrayList<Int>>): ArrayList<ArrayList<Int>> {
    var previous: ArrayList<ArrayList<Int>> = ArrayList(matrix.size)
    for( i in 0..matrix.size - 1)
        previous.add(ArrayList<Int>())

    for(i in 0..matrix.size - 1)
        for(j in 0..matrix.size - 1)
            if (matrix.get(i).get(j) > 0)
                previous.get(j).add(i)
    return previous
}

//Вычисление ранних сроков свершения событий
private fun calcualte_Tp(matrix: ArrayList<ArrayList<Int>>, prev: ArrayList<ArrayList<Int>>): ArrayList<Int>{
    var Tp: ArrayList<Int> = ArrayList()
    for(i in 0..matrix.size - 1)
        Tp.add(0)

    for(i in 0..Tp.size - 1){
        var previous: ArrayList<Int> = prev.get(i)
        var temp: ArrayList<Int> = ArrayList()
        for(j in previous){
            if(previous.size == 1 && j == 0)
                Tp.set(i, matrix.get(j).get(i))
            else
                temp.add(matrix.get(j).get(i) + Tp.get(j))
        }
        if(temp.size != 0)
            Tp.set(i, temp.maxOrNull() ?: 0)
    }
    return Tp
}

//Вычисление поздних сроков свершения событий
private fun calculate_Tr(matrix: ArrayList<ArrayList<Int>>, Tp: ArrayList<Int>): ArrayList<Int> {
    var Tr: ArrayList<Int> = ArrayList()
    for(i in 0..matrix.size - 1)
        Tr.add(0)
    Tr.set(Tp.size - 1, Tp.get(Tp.size - 1))
    for(i in Tp.size - 2 downTo 0){
        var temp: ArrayList<Int> = ArrayList()
        for(j in 0..Tr.size - 1)
            if(matrix.get(i).get(j) > 0)
                temp.add(Tr.get(j) - matrix.get(i).get(j))
        Tr.set(i, temp.minOrNull() ?: 0)
    }
    return Tr
}

//Вычисление критического пути
private fun calculate_critical_path(Tp: ArrayList<Int>, Tr: ArrayList<Int>): ArrayList<Int>{
    var critical_path: ArrayList<Int> = ArrayList()
    for(i in 0..Tr.size - 1)
        if(Tr.get(i) == Tp.get(i))
            critical_path.add(i)
    return critical_path
}

//Проверка размера матрицы (размер должен быть n x n)
private fun check_matrix(matrix: ArrayList<ArrayList<Int>>): Boolean{
    for(i in 0..matrix.size - 1)
        if(matrix.size == matrix.get(i).size)
            continue
        else
            return false
    return true
}

//Вычисление всех резервов времени работы
private fun calculate_all_reserve(matrix: ArrayList<ArrayList<Int>>, Tp: ArrayList<Int>, Tr: ArrayList<Int>): ArrayList<HashMap<String, Int>>{
    var reserve: ArrayList<HashMap<String, Int>> = ArrayList()
    var Rfull: HashMap<String, Int> = HashMap()
    var Rguaranteed: HashMap<String, Int> = HashMap()
    var Rfree: HashMap<String, Int> = HashMap()
    var Rindependent: HashMap<String, Int> = HashMap()

    for(i in 0..matrix.size - 1)
        for(j in 0..matrix.size - 1)
            if(matrix.get(i).get(j) > 0){
                Rfull.put(i.toString() + "-" + j.toString(),  Tr.get(j) - Tp.get(i) - matrix.get(i).get(j))
                Rguaranteed.put(i.toString() + "-" + j.toString(), Tr.get(j) - Tr.get(i) - matrix.get(i).get(j))
                Rfree.put(i.toString() + "-" + j.toString(), Tp.get(j) - Tp.get(i) - matrix.get(i).get(j))
                Rindependent.put(i.toString() + "-" + j.toString(), Tp.get(j) - Tr.get(i) - matrix.get(i).get(j))
            }
    reserve.add(Rfull)
    reserve.add(Rguaranteed)
    reserve.add(Rfree)
    reserve.add(Rindependent)
    return reserve
}

//Вывод ранних и поздних сроков свершения событий
private fun print_Tp_Tr(Tp: ArrayList<Int>, Tr: ArrayList<Int>){
    print("_____________Сроки свершения событий______________")
    println()
    print(String.format("%-13s", "Событие"))
    print(String.format("%-13s", "Ранние"))
    print(String.format("%-13s", "Поздние"))
    print(String.format("%-13s", "Резерв"))
    println()
    for(i in 0..Tp.size - 1){
        print(String.format("%-13s", i))
        print(String.format("%-13s", Tp[i]))
        print(String.format("%-13s", Tr[i]))
        print(String.format("%-13s", Tr[i] - Tp[i]))
        println()
    }
    println()
}

//Вывод маршрута критического пути
private fun print_critical_path(critical_path: ArrayList<Int>){
    print("Критический путь: ")
    for(i in 0..critical_path.size - 1){
        if(i < critical_path.size - 1)
            print(critical_path.get(i).toString() + " -> ")
        else
            print(critical_path.get(i))
    }
    println()
}

//Вывод резервов работы
private fun print_reserve(reserve: ArrayList<HashMap<String, Int>>){
    
    println("_______________________________Резервы работ_________________________________")
    print(String.format("%-13s", "Связь"))
    print(String.format("%-13s", "Полный"))
    print(String.format("%-18s", "Гарантированный"))
    print(String.format("%-18s", "Свободный"))
    print(String.format("%-16s", "Независимый"))
    println()

    for(i in reserve[0].keys){
        print(String.format("%-13s", i))
        print(String.format("%-13s", reserve[0][i]))
        print(String.format("%-18s", reserve[1][i]))
        print(String.format("%-18s", reserve[2][i]))
        print(String.format("%-13s", reserve[3][i]))
        println()
    }
}

//Вычисление параметров сетевой модели
public fun calcualte_parameters(matrix: ArrayList<ArrayList<Int>>){
    if(check_matrix(matrix)){
        var previous: ArrayList<ArrayList<Int>> = calcualte_previous(matrix)

        var Tp: ArrayList<Int> = calcualte_Tp(matrix, previous)
        var Tr: ArrayList<Int> = calculate_Tr(matrix, Tp)
        print_Tp_Tr(Tp, Tr)
        
        var reserve: ArrayList<HashMap<String, Int>> = calculate_all_reserve(matrix, Tp, Tr)
        print_reserve(reserve)

        var critical_path: ArrayList<Int> = calculate_critical_path(Tp, Tr)
        print_critical_path(critical_path)
    }
    else
        println("Неверный размер матрицы!")
}

File: prog_golang/network_plann/network_plann.go
package network_plann

import (
	"fmt"
	"os"
	"strconv"
)

// Вычисление предшественников
func calculate_previous(matrix *[][]int) [][]int {
	var previous = make([][]int, len(*matrix))
	for i, n_i := range *matrix {
		for j, n_j := range n_i {
			if n_j > 0 {
				previous[j] = append(previous[j], i)
			}
		}
	}
	for i := 0; i < len(previous); i++ {
		for j := 0; j < len(previous[i]); j++ {
		}
	}
	return previous
}

// Вычисление ранних сроков свершения событий
func calculate_Tp(matrix *[][]int, prev *[][]int) []int {
	var Tp = make([]int, len(*matrix))
	for i := 0; i < len(*matrix); i++ {
		var previous []int = (*prev)[i]
		var temp []int
		for _, j := range previous {
			if len(previous) == 1 && j == 0 {
				Tp[i] = (*matrix)[j][i]
			} else {
				temp = append(temp, (*matrix)[j][i]+Tp[j])
			}
		}
		if len(temp) != 0 {
			max_element := 0
			for k := 0; k < len(temp); k++ {
				if max_element < temp[k] {
					max_element = temp[k]
				}
			}
			Tp[i] = max_element
		}
	}
	return Tp
}

// Вычисление поздних сроков свершения событий
func calculate_Tr(matrix *[][]int, Tp *[]int) []int {
	var Tr = make([]int, len(*matrix))
	Tr[len(*Tp)-1] = (*Tp)[len(*Tp)-1]
	for i := len(Tr) - 2; i >= 0; i-- {
		var temp []int
		for j := 0; j < len(Tr); j++ {
			if (*matrix)[i][j] > 0 {
				temp = append(temp, Tr[j]-(*matrix)[i][j])
			}
		}
		if len(temp) != 0 {
			min_element := temp[0]
			for k := 0; k < len(temp); k++ {
				if min_element > temp[k] {
					min_element = temp[k]
				}
			}
			Tr[i] = min_element
		}
	}
	return Tr
}

// Вычисление критического пути
func calculate_critical_path(Tr *[]int, Tp *[]int) []int {
	var critical_path []int
	for i := 0; i < len(*Tr); i++ {
		if (*Tr)[i] == (*Tp)[i] {
			critical_path = append(critical_path, i)
		}
	}
	return critical_path
}

// Проверка размера матрицы (размер должен быть n x n)
func check_matrix(matrix *[][]int) bool {
	for i := 0; i < len(*matrix); i++ {
		if len(*matrix) == len((*matrix)[i]) {
			continue
		} else {
			return false
		}
	}
	return true
}

// Вычисление всех резервов времени работы
func calculate_all_reserve(Tr *[]int, Tp *[]int, matrix *[][]int) (map[string]int, map[string]int, map[string]int, map[string]int) {
	Rfull := make(map[string]int)        // Полный резерв времени работы
	Rguaranteed := make(map[string]int)  // Гарантированный резерв времени работы
	Rfree := make(map[string]int)        // Свободный резерв времени работы
	Rindependent := make(map[string]int) //# Независимый резерв времени работы

	for i := 0; i < len(*matrix); i++ {
		for j := 0; j < len(*matrix); j++ {
			if (*matrix)[i][j] > 0 {
				Rfull[strconv.Itoa(i)+"-"+strconv.Itoa(j)] = (*Tr)[j] - (*Tp)[i] - (*matrix)[i][j]
				Rguaranteed[strconv.Itoa(i)+"-"+strconv.Itoa(j)] = (*Tr)[j] - (*Tr)[i] - (*matrix)[i][j]
				Rfree[strconv.Itoa(i)+"-"+strconv.Itoa(j)] = (*Tp)[j] - (*Tp)[i] - (*matrix)[i][j]
				Rindependent[strconv.Itoa(i)+"-"+strconv.Itoa(j)] = (*Tp)[j] - (*Tr)[i] - (*matrix)[i][j]
			}
		}
	}
	return Rfull, Rguaranteed, Rfree, Rindependent
}

// Вывод ранних и поздних сроков свершения событий
func print_Tp_Tr(Tp *[]int, Tr *[]int) {
	fmt.Println("________________Сроки свершения событий_________________")
	fmt.Fprintf(os.Stdout, "%-15s %-15s %-15s %-15s", "Событие", "Ранние", "Поздние", "Резерв")
	fmt.Println()
	for i := 0; i < len(*Tr); i++ {
		fmt.Fprintf(os.Stdout, "%-15d %-15d %-15d %-15d", i, ((*Tp)[i]), ((*Tr)[i]), ((*Tr)[i] - (*Tp)[i]))
		fmt.Println()
	}
	fmt.Print("\n\n")
}

// Вывод маршрута критического пути
func print_critical_path(critical_path *[]int) {
	fmt.Println("Критический путь:")
	for i := 0; i < len(*critical_path); i++ {
		if i != len(*critical_path)-1 {
			fmt.Print((*critical_path)[i], " -> ")
		} else {
			fmt.Print((*critical_path)[i], "\n\n")
		}
	}
}

// Вывод резервов работы
func print_reserve(Rfull *map[string]int, Rguaranteed *map[string]int, Rfree *map[string]int, Rindependent *map[string]int) {
	fmt.Println("_____________________________Резервы работ_____________________________")
	fmt.Fprintf(os.Stdout, "%-10s %-10s %-20s %-15s %-15s", "Связь", "Полный", "Гарантированный", "Свободный", "Независимый")
	fmt.Println()
	for key, _ := range *Rfull {
		fmt.Fprintf(os.Stdout, "%-10s %-10d %-20d %-15d %-10d", key, (*Rfull)[key], (*Rguaranteed)[key], (*Rfree)[key], (*Rindependent)[key])
		fmt.Println()
	}
}

// Вычисление параметров сетевой модели
func Calculate_parameters(matrix *[][]int) {
	if check_matrix(matrix) {
		ptr_matrix := matrix
		previous := calculate_previous(ptr_matrix)
		ptr_previous := &previous

		Tp := calculate_Tp(ptr_matrix, ptr_previous)
		ptr_Tp := &Tp
		Tr := calculate_Tr(ptr_matrix, ptr_Tp)
		ptr_Tr := &Tr
		print_Tp_Tr(ptr_Tp, ptr_Tr)

		Rfull, Rguaranteed, Rfree, Rindependent := calculate_all_reserve(ptr_Tr, ptr_Tp, ptr_matrix)
		ptr_rfull := &Rfull
		ptr_rguaranteed := &Rguaranteed
		ptr_rfree := &Rfree
		ptr_rindependent := &Rindependent
		print_reserve(ptr_rfull, ptr_rguaranteed, ptr_rfree, ptr_rindependent)

		critical_path := calculate_critical_path(ptr_Tr, ptr_Tp)
		prt_critical_path := &critical_path
		print_critical_path(prt_critical_path)
	} else {
		fmt.Println("Неверный размер матрицы!")
	}
}
